2020-11-26 unit testing

punkyiang opened the pull request:
For now I've added unit tests for the wallet deletion and transfer functionality.
Another thing that I did is that I forgot to commit the changes made at the lesson and separate them from the homework, thus the code from the lesson is still there. I've tried to stash the new changes and commit the lesson stuff separately, but the stash functionality just disappeared from Visual Studio 16.8.2 with its new Git Changes window (or at least I cannot find it). Staging just part of the scripts wouldn't work as parts of the old scripts were removed (such as the old static CurrencyManager, now it is an actual injectable interface and stuff). Is there a way to properly rearrange it all into a couple commits so that I don't have to manually copy/paste the changes in the notepad?

punkyiang committed:
2020-11-26 - unit testing

TermanEmil commented on PaymentSystem.Server.UnitTests/CreateWalletCommandHandlerTests.cs:
```diff
@@ -0,0 +1,127 @@
+        public async Task CreateWalletSuccessful()
+        {
+
+            var sut = new CreateWalletCommandHandler(context, promotionManagerMock.Object, currencyManagerMock.Object);
```
you could initialize `sut` in setup, no?

punkyiang replied:
Thanks for the tip! I've rewritten the code so the systems under test are initialized in the SetUp method

TermanEmil commented on PaymentSystem.Server.UnitTests/CreateWalletCommandHandlerTests.cs:
```diff
@@ -0,0 +1,127 @@
+{
+    public class CreateWalletCommandHandlerTests
+    {
+        private ApplicationDbContext context;
```
Isn't `ApplicationDbContext` implementing `IDisposable`? If so, it must be called.
Just like there's a SetUp, there's TearDown where you can call `context.Dispose()`

punkyiang replied:
It is indeed! Thanks for the heads-up! Fixed it in the newer commit

TermanEmil commented on PaymentSystem.Server.UnitTests/MakeTransferCommandHandlerTests.cs:
```diff
@@ -0,0 +1,198 @@
+
+            var result = await sut.Handle(command, CancellationToken.None);
+
+            Assert.IsTrue(result.IsSuccessful);
```
Perhaps we could query the db if a transfer was actually made.

punkyiang replied:
Good idea! I've added db queries in places where it's relevant. Thanks!

TermanEmil commented on PaymentSystem/Server/Application/Currencies/CurrencyManager.cs:
```diff
@@ -3,22 +3,28 @@
             };
         }
+
+        public List<string> GetCurrencies()
```
I think it would be nice to keep the Currencies as a property.
```csharp
public IEnumerable<string> Currencies { get; }
```
`IEnumerable` so we're not allowing others to modify the currencies.

punkyiang replied:
Interesting thought, this might reduce the chance for future problems, thanks!

TermanEmil commented on PaymentSystem/Server/Application/Transfers/Commands/MakeTransferCommand.cs:
```diff
@@ -12,27 +12,58 @@
             if (!user.Wallets.Any(x => x.Currency == command.Data.Currency))
             {
-                return BoolResult.ReturnFailure();
+                return MakeTransferResult.ReturnFailure("MISSING_SOURCE_WALLET");
```
Should we have constants for this somewhere?

punkyiang replied:
Yep, I've added internal response message enums for each result, thus keeping the code more readable, thanks!

punkyiang committed:
refactored unit testing
- set command/query execution messages as enums
- added database queries as additional tests
- set currencies as IEnumerable<string>, so one would not be able to edit them
- added TearDown methods which dispose of the ApplicationDbContext
